# Mathematics

# The Fundamentals Series

## Section Four : Fractions

- 4.01 Simplification & Cancellation
- 4.02 Equivalent fractions
- 4.03 Division of zero
- 4.04 Addition & Subtraction of fractions
- 4.05 Multiplication & Division of fractions
- 4.06 Signs of fractions
- 4.07 Reciprocals

Fractions are rational numbers which means they ratio of integers. Fractions should be thought of as part of a whole. a fraction will always be a two part structure where we have the numerator and the denominator. Typically it will physically resemble $\frac {2}{3}$ or $\frac {11}{17}$ etc.. Fractional structures can be looked at as demonstrating the partial value between zero and one. 

In [134]:
# python imports
from collections import Counter
import numpy as np

## 4.01 Simplification & Cancellation

The aim is to reduce a fraction to its lowest terms. This is done by simplification and cancellation, for instance, assume we look at $\frac {50}{100}$ we can see that this is a half, which is the same as $\frac {5}{10}$. But we can further simplify. Indeed it's the same as $\frac {2}{4}$ but that's still not the lowest terms. We can see that a half or 50% can be written as $\frac {1}{2}$ and this is the lowest terms we can reach for our values.  

To deduce the lowest terms we can use prime factorization again. 

In [135]:
# returns the product values as a list. 
def product_of_primes(num):
    
    # produce a list comprehension of all prime numbers between 2-500
    primes = [x for x in range(2, 500) if all(x % y != 0 for y in range(2, x))]
    
    # create a list for our products 
    pr = []
    working_num = num
    
    # run the list 
    for prime in primes:
        if working_num % prime == 0:
            while working_num % prime == 0:
                working_num /= prime
                # factor identified, add to the list 
                pr.append(prime)
                
    # print the completed list
    return pr

#### Example 1

Problem Statement Definition: $\frac {10}{24}$

In [136]:
# get our product of primes values.
product_of_primes(10), product_of_primes(24)

([2, 5], [2, 2, 2, 3])

We can now use cancellation: 

- Identify cancellation targets as the leading 2 from both the numerator and denominator. 

- We can therefore express the problem as: $\frac {10}{24}$ as $\frac {5}{2^2 \cdot 3}$ This may still be confusing to look at therefore we can resolve $2^2 \cdot 3$ as $2 \cdot 2 \cdot 3$ or 12.  

- The complete reduction process gives us $\frac {5}{12}$

#### Example 2

Problem statement definition: $\frac {54}{81}$

In [137]:
product_of_primes(54), product_of_primes(81)

([2, 3, 3, 3], [3, 3, 3, 3])

Process of cancellation details: 

- allows us to remove three 3's from each side of our fraction, thus leaving us with $\frac {2}{3}$

#### Example 3

Problem statement definition: $\frac {56}{27} \cdot \frac{14}{21} \cdot \frac{15}{22}$

In [138]:
# step 1 - calculate the product of primes for 56/27
product_of_primes(56), product_of_primes(27)

([2, 2, 2, 7], [3, 3, 3])

In [139]:
# step 2 - calculate the product of primes for 14/21
product_of_primes(14), product_of_primes(21)

([2, 7], [3, 7])

In [140]:
# step 3 - calculate the product of primes for 15/22
product_of_primes(15), product_of_primes(22)

([3, 5], [2, 11])

At this point we now have: $\frac{2^3 \cdot 7}{3^3} \cdot \frac{2}{3} \cdot \frac{3 \cdot 5}{2 \cdot 11}$

Now we look to cancellation:

- Identifying we have 3 in the denominator in $\frac{2}{3}$ and 15 in the numerator of $\frac{15}{22}$ and the 2 and 11 are also divisible, therefore that leaves $\frac{1}{1} \cdot \frac{5}{11}$ meaning the $\frac{1}{1}$ can be canceled completely.
- Now we have: $\frac{2^3 \cdot 7}{3^3} \cdot \frac{5}{11}$
    
We cannot reduce any further, so we now have to take the whole numerator/denominator result. $56 \cdot 5$ and $27 \cdot 11$ = $\frac{280}{297}$

#### Example 4

Problem statement Definition: $\frac{12}{105} \cdot \frac{14}{30} \cdot \frac{15}{14}$

- Immediate cancellation allows us to see the matches in the last two fractions, the 14's in the numerator and demonimator cancel each other, the 15 and 30 can be reduced, therefore we have $\frac{1}{2} \cdot \frac{1}{1}$ the last of which can be cancelled completely. Leaving the new problem statement to be: $\frac{12}{105} \cdot \frac{1}{2}$
- Further cancellation, we can see the 12 in the numerator and the 2 in the denominator are divisibe, leaving:$\frac{6}{105} \cdot \frac{1}{1}$ and the $\frac{1}{1}$ can be removed completely. We now have $\frac{6}{105}$ which are divisible by 3. 
- After the divison by 3 we have $\frac{2}{35}$

#### Example 5

Problem statement definition $\frac{50}{85}$

In [141]:
# get the product of primes 
product_of_primes(50), product_of_primes(85)

([2, 5, 5], [5, 17])

cancellation steps 

- Removing common 5's leaving $\frac{2 \cdot 5}{17}$ or $\frac{10}{17}$

## 4.02 Equivalent Fractions

We have seen how to get to the lowest terms above, now we can look at equivalence. Equivalence is expressing the same fraction with a different base. For example we can express $\frac{2}{3}$ as $\frac{4}{6}$ or $\frac{6}{9}$ or even $\frac{70}{105}$ They all represent the same share of the whole.   

In the opposite direction with reduction we can say that $\frac{80}{100}$ is equivalent to $\frac{4}{5}$

## 4.03 Division of Zero

Zero in the denominator means a fraction is "undefined". Division of zero is not possible and therefore not allowed, zero in the numerator means zero, zero in the denominator is not a possible operation.


## 4.04 Addition & Subtraction of Fractions

Before addition of subtraction of fractions is possible we must achieve the same denominator. Where the denominator is already the same the operation can proceed, where the denominator is different we must find the LCM (Least, or lowest, common multiple)

In [142]:
# We have seen the LCM before in the factors
# and multiples notebook. Therefore the repeat 
# of a concept means that we need a lowest 
# common multiple function for the purpose of 
# use here and future use. 

In [143]:
def product_of_primes(num):
    
    # produce a list comprehension of all prime numbers between 2-500
    primes = [x for x in range(2, 500) if all(x % y != 0 for y in range(2, x))]
    
    # create a list for our products 
    pr = []
    working_num = num
    
    # run the list 
    for prime in primes:
        if working_num % prime == 0:
            while working_num % prime == 0:
                working_num /= prime
                # factor identified, add to the list 
                pr.append(prime)
                
    # print the completed list
    return pr

In [144]:
def lowest_common_multiple(*args):
    monomials = [arg for arg in args]
    lcm_factors = []
    
    for mono in monomials:
        elements = dict(Counter(element for element in product_of_primes(mono)))
        
        for k, v in elements.items():
            if v > 1:
                res = k ** v    
            else:
                res = k
            
            if lcm_factors.count(res) == 0:
                    lcm_factors.append(res)
                
    return np.prod(lcm_factors)
    

In [145]:
# Check the product pf primes structures 
product_of_primes(40), product_of_primes(65) 

([2, 2, 2, 5], [5, 13])

In [146]:
# get the least common multiple 
res = lowest_common_multiple(40, 65)
res

520

In [147]:
# Take the divisor as the multiplier for the numerator value 
(res / 65) * 23

184.0

In [148]:
# Take the divisor as the multiplier for the numerator value 
(res / 40)  * 17

221.0

In [149]:
184 + 221

405

Step result of our operations is $\frac {405}{520}$

In [150]:

# creates a list of all matching divisors for a group of
# monomials 

def greatest_common_factor(*args):
    monomials = [arg for arg in args]
    divisors = []
    t_minus = max(monomials)
    lower_bound = 1
    
    while t_minus >= lower_bound:
        is_divisor = True
        for mono in monomials:
            if mono % t_minus != 0:
                is_divisor = False
                
        if is_divisor:
            divisors.append(t_minus)
       
        t_minus -= 1
    return divisors[0]

In [151]:
# we can observe that as the numbers end in 5 and 0 that 5 will be a prime divisor
# hwever to check that we can run the greatest common factor funtion. Returning 5.
res = greatest_common_factor(405, 520)
res

5

In [152]:
405 / 5, 520/5 

(81.0, 104.0)

final result is: $\frac{81}{104}$

#### Exercise 1

Simplify the expression $\frac {1}{5} + \frac{3}{7}$

In [153]:
res=lowest_common_multiple(5, 7)
res

35

Step result:  $\frac{7}{35} + \frac{15}{35}$ =  $\frac{22}{35}$

#### Exercise 2

What is the least common denominator of the fractions: $\frac{1}{2} + \frac{1}{8} + \frac{1}{20}$

In [154]:
res = lowest_common_multiple(2, 8, 20)
res

320

The least common multiple is 320. 

step result is: $\frac{160}{320} + \frac{40}{320} + \frac{16}{320} = \frac{216}{320}$

Identified 2 as a common divisor. $\frac{108}{160}$ then $\frac{54}{80}$ then $\frac{27}{40}$ 

#### Exercise 3

Simplify the expression: $\frac{10}{7} - \frac{6}{15}$ 

In [155]:
# step one, find the least common multiple 
res = lowest_common_multiple(7, 15)
res

105

In [156]:
# get the totals for the numerators with the appropriate multipliers.
10 * (105 / 7), 6 * (105 / 15)

(150.0, 42.0)

Yields: $\frac{108}{105}$

In [157]:
# grab the greatest common factor 
res = greatest_common_factor(108, 105)
res

3

In [158]:
# reduce the fraction to the lowest terms by applying the 
# division by the greatest common factor. 
108 / 3, 105 /3

(36.0, 35.0)

In [159]:
# Double check it cannot be reduced further. 
# an answr of 1 conforms we are at the lowest terms 
res = greatest_common_factor(16, 35)
res

1

## 4.05 Multiplication & Division of Fractions

- Multiplication:
    - Multiply the numerators together
    - Multiple the denominators together
    - reduce the product to the lowest terms
    
- Division:
    - Division examples should be changed to a multiplication example instead, we do this by finding the reciprocal. 
    - To get the reciprocal, we flip the positions of the numerator and denominator of the 2nd fraction
    - The multiplication is performed and reduced to lowest terms. 

Additional note: 
Division examples may be written in two equally common forms.  $\frac{5}{8} \div \frac{1}{2}$  or $\frac{\frac{5}{8}}{\frac{1}{2}}$

#### Exercise 1

Simplify the expression: $\frac{2}{21} \cdot \frac{3}{5}$ 

In [160]:
# perform the cross multiplications

2 * 3, 21 * 5

(6, 105)

In [161]:
res = greatest_common_factor(6, 105)
res

3

Yields:  $\frac{2}{35}$ 

#### Example 2

Multiply the fractions: $\frac{3}{5} \cdot \frac{1}{2} \cdot \frac{5}{6}$



In [162]:
3 * 1 * 5, 5 * 2 * 6

(15, 60)

$\frac{15}{60}$ can be reduced to $\frac{3}{12}$ which in turn can be reduced by 3, yielding $\frac{1}{4}$

#### Example 3

resolve $\frac{1}{2} \div \frac{1}{7}$

convert to a multiplication problem using the reciprocal of the 2nd fraction.

we now have $\frac{1}{2} \cdot \frac{7}{1}$

In [163]:
1 * 7, 2 * 1

(7, 2)

result is:  $\frac{7}{2}$

## 4.06 Signs of Fractions

Up until now we have looked at and covered only positive fractions. Fractions can have 3 signs possibilities: 
- $\frac{-2}{3}$

- $\frac{2}{-3}$

- $- \frac{2}{3}$

Any or all of these can be a factor in our problem statement. To simplify the problem statement we look to match the signs of fractions where they differ. 

- You can change any two signs without changing the value of the statement. This means: $+\frac{+1}{-2}$ is the same as $- \frac{1}{2}$

#### Example 1

$\frac{+1}{x -3} - \frac{+4}{-x +3}$

Step one, we want to get the signs of the denominators to match, on the left we have a positive `x` and on the right we have a negative `x`. To do so we can parenthesise the right denominator that implies a positive result: $\frac{+1}{x -3} - \frac{+4}{(-x +3)}$ instead we can flip that to be negative: $\frac{+1}{x -3} - \frac{+4}{-(-x +3)}$  We can now cancel the double negative to a positive and flip the remaining signs giving us: $\frac{1}{x -3} + \frac{4}{x -3}$ and finally $\frac{5}{x -3}$

## 4.07 Reciprocals

The reciprocal of a fraction is the fraction with the numerator and denominator flipped. ie. $\frac{1}{3}$ becomes $\frac{3}{1}$ If a negative sign is present then we must carry that also, meaning: $\frac{-9}{2}$ becomes $\frac{2}{-9}$

If the entire fraction is a negative that should persist. Thus: $- \frac{7}{3}$ becomes $-\frac{3}{7}$

Notes: 
- if you multiply a fraction by it's reciprocal you will always yield 1. 
- Even if the entire fraction is negative, a reciprocal multiplication will yield 1, with sign cancellation applying. 
- The purpose is to answer what we must apply to achieve 1. 
- Therefore the only number without a reciprocal is 0. 